File: python/stftpitchshift/vocoder.py
import numpy as np


def wrap(x):

    return (x + np.pi) % (2 * np.pi) - np.pi


def encode(frames, framesize, hopsize, samplerate):

    # return encode_orig(frames, framesize, hopsize, samplerate)
    return encode_tfr(frames, framesize, hopsize, samplerate)


def encode_tfr(frames, framesize, hopsize, samplerate):

    return frames


def encode_orig(frames, framesize, hopsize, samplerate):

    M, N = frames.shape

    analysis_framesize = np.ravel(framesize)[0]
    synthesis_framesize = np.ravel(framesize)[-1]

    freqinc = samplerate / analysis_framesize
    phaseinc = 2 * np.pi * hopsize / analysis_framesize

    buffer = np.zeros(N)
    data = np.zeros((M, N), complex)

    for m, frame in enumerate(frames):

        abs = np.abs(frame)
        arg = np.angle(frame)

        delta = arg - buffer
        buffer = arg

        i = np.arange(N)
        j = wrap(delta - i * phaseinc) / phaseinc

        freq = (i + j) * freqinc

        data[m] = abs + 1j * freq

    return data


def decode(frames, framesize, hopsize, samplerate):

    M, N = frames.shape

    analysis_framesize = np.ravel(framesize)[0]
    synthesis_framesize = np.ravel(framesize)[-1]

    freqinc = samplerate / analysis_framesize
    phaseinc = 2 * np.pi * hopsize / analysis_framesize

    # compensate asymmetric synthesis window by virtual time shifting #38
    timeshift = 2 * np.pi * synthesis_framesize * np.arange(N) / N \
        if synthesis_framesize != analysis_framesize else 0

    buffer = np.zeros(N)
    data = np.zeros((M, N), complex)

    for m, frame in enumerate(frames):

        abs = np.real(frame)
        freq = np.imag(frame)

        i = np.arange(N)
        j = (freq - i * freqinc) / freqinc

        delta = (i + j) * phaseinc

        buffer += delta
        arg = buffer.copy()

        arg -= timeshift #38

        data[m] = abs * np.exp(1j * arg)

    return data

File: cpp/StftPitchShift/Vocoder.cpp
#include <StftPitchShift/Vocoder.h>

Vocoder::Vocoder(const size_t framesize, const size_t hopsize, const float samplerate) :
  stft_freq_inc((double)samplerate / (double)framesize),
  stft_phase_inc(PI2 / ((double)framesize / (double)hopsize))
{
  encode_phase_buffer.resize(framesize / 2 + 1);
  decode_phase_buffer.resize(framesize / 2 + 1);
}

void Vocoder::encode(std::vector<std::complex<float>>& dft)
{
  double frequency,
         phase,
         delta,
         j;

  for (size_t i = 0; i < dft.size(); ++i)
  {
    phase = std::arg(dft[i]);

    delta = phase - encode_phase_buffer[i];
    encode_phase_buffer[i] = phase;

    j = wrap(delta - i * stft_phase_inc) / stft_phase_inc;

    frequency = (i + j) * stft_freq_inc;

    dft[i] = std::complex<float>(
      std::abs(dft[i]),
      static_cast<float>(frequency));
  }
}

void Vocoder::decode(std::vector<std::complex<float>>& dft)
{
  double frequency,
         phase,
         delta,
         j;

  for (size_t i = 0; i < dft.size(); ++i)
  {
    frequency = dft[i].imag();

    j = (frequency - i * stft_freq_inc) / stft_freq_inc;

    delta = (i + j) * stft_phase_inc;

    decode_phase_buffer[i] += delta;
    phase = decode_phase_buffer[i];

    dft[i] = std::polar<float>(
      dft[i].real(),
      static_cast<float>(phase));
  }
}
